### PR TITLE
Add Kollider Lite Umbrel app

### PR DESCRIPTION
**Pull request by djkazic**
# App Submission

### App name
Kollider

### Version
1.0.0

### One line description of the app
_(max 50 characters)_

Lightning-Fast Derivative Trading

...

### Summary of the app
_(50 to 200 words)_

Kollider lets you instantly trade perpetual contracts with low fees and up to 100x buying power. 
No need to pre-fund a trading account, each trade settles directly from your own wallet or even from your own lightning node in milliseconds. 

The Kollider Umbrel App lets you use the Lightning Network to get instant exposure to a range of products using Bitcoin, directly from your Umbrel node.

More features coming soon! 

...

### Developer name
Kollider

### Developer website
https://www.pro.kollider.xyz

### Source code link
https://github.com/kolliderhq/kollider-lite-app

...

### Support link
Telegram: https://t.me/kolliderhq

...

### Requires
- [ ] Bitcoin Core
- [ ] Electrum server
- [X] LND

### 256x256 SVG icon

https://svgshare.com/s/diu

### Gallery images
_(Upload 3 to 5 high-quality gallery images (1440x900px) of your app in PNG format, or just upload 3 to 5 screenshots of your app and we'll help you design the gallery images.)_

Lightning-Fast Derivative Trading.
Lightning-Fast Derivative Trading.
Instantly open and close multiple positions directly from your own node.
Instantly open and close multiple positions directly from your own node.
Compete and stack sats against other traders over the Lightning Network.
Compete and stack sats against other traders over the Lightning Network.

### I have tested my app on:
- [X] [Umbrel dev environment](https://github.com/getumbrel/umbrel-dev)
- [ ] [Umbrel OS on a Raspberry Pi 4](https://github.com/getumbrel/umbrel-os)
- [X] [Custom Umbrel install on Linux](https://github.com/getumbrel/umbrel#-installation)

### CODE REVIEW

**Review by lukechilds — CHANGES_REQUESTED** [review]
Kollider looks really cool, nice work guys!

Looks there are a few bits needed to get this working. Added some comments with some questions. I also fixed a few bits in this PR: https://github.com/kolliderhq/umbrel/pull/1

For future reference, it makes PR review a bit easier if you submit PRs from a feature branch and check "Allow edits from maintainers" so I can push changes directly to your PR branch. No worries though, you can just merge the above PR  for now!

Let me know if any of the questions are unclear.

File: apps/kollider/docker-compose.yml
Comment (djkazic) [reply apps/kollider/docker-compose.yml:5]: Cool, we'll get those arm builds out ASAP.

File: apps/kollider/docker-compose.yml
Comment (djkazic) [reply apps/kollider/docker-compose.yml:24]: We're currently using the static defined LND IP to connect. Would that change in the future?

**Comment by lukechilds** [comment]
Also guys would you be able to give me a breif description of what the three container do and how they relate to eachother?

File: apps/kollider/docker-compose.yml
Comment (djkazic) [reply apps/kollider/docker-compose.yml:22]: I'll get this fixed now, thanks for the tip. This is serving a websocket server so the frontend can connect to it.

File: apps/kollider/docker-compose.yml
Comment (lukechilds) [reply apps/kollider/docker-compose.yml:24]: Yeah those will be assigned dynamically at some point in the future and not hardcoded so each Umbrel will have different IPs. You can pass in the connection details based on the env vars documented here: https://github.com/getumbrel/umbrel/tree/master/apps#2-%EF%B8%8Fpackaging-the-app-for-umbrel

These are the ones available for LND:
```
      # LND environment variables
      # $LND_IP - Local IP of LND
      # $LND_GRPC_PORT - gRPC Port of LND
      # $LND_REST_PORT - REST Port of LND
```

**Comment by djkazic** [comment]
Yep
1. kollider-lite-app is the front-end itself
2. ws-client is a websocket server that connects over zmq to the backend, which actually talks to the kollider api
3. backend is a python server that bridges ws-client to our api

**Comment by lukechilds** [comment]
Got it, thanks!

**Comment by lukechilds** [comment]
Also not sure if you missed it but there are a few fixes in here: https://github.com/kolliderhq/umbrel/pull/1

**Comment by djkazic** [comment]
Thanks, I'll incorporate those changes

**Comment by djkazic** [comment]
@lukechilds I think I've got the changes requested, do you mind having another look 👀

**Comment by lukechilds** [comment]
Of course! The UI now loads for me but it doesn't connect to your remote API due to a missing CORS header, it also doens't seem to connect back to the local WebSocket server.

Screenshot 2022-01-30 at 18 15 57

How does the frontend know where the ws server is running and where to connect to? I don't see any env vars passed through to the frontend defining that. Is it expecting the ws server to be running on the same host on a hardcoded port? I don't see any attempts to connect to any kind of ws server in dev tools, there is an unhandled error from NextJS so maybe something is going wrong before that happens?

Also the LND data dir is passed into both the `ws` and `backend` service but the LND IP is only passed into the `backend` service. Are both services connecting to LND?

I also noticed you're passing in `$APP_PASSWORD` but I'm able to see the UI without being prompted for a password.

**Comment by djkazic** [comment]
Hey @lukechilds, in order to make the app work you've got to go to umbrel.local:4243. Thats the domain that's whitelisted on our API. It's the same behavior as the lnmarkets app.

Only backend connects to lnd, the ws talks to the backend via ZMQ.

The frontend auth only pops up when there is a valid connection to the ws.

**Comment by djkazic** [comment]
image
This is how it should look when you access it from the `umbrel.local:4243` URL.

**Comment by lukechilds** [comment]
>Hey @lukechilds, in order to make the app work you've got to go to umbrel.local:4243. Thats the domain that's whitelisted on our API. 

Ahh got it! Ok changed my hostname to umbrel.local and it's working great! Man I've gotta admit this is suuuper slick, really amazing work!

One issue is this app won't work when accessed remotely over Tor. Since it'll be served from `${random_chars}.onion` not `umbrel.local`. But that's no big deal, we can disable it for remote Tor users for now. And hopefully enable it in the future if you can handle alternative hostnames.

One other thing I noticed, you're loading 3rd party JS from https://plausible.io/js/plausible.js. This is a bit of an issue since it's essentially a back door into your app. If that domain ever served up bad JS it would allow them to take control of the app and open positions without the user's consent. Looks like it can be solved by simply inlining that JS into your app.

File: apps/registry.json
Comment (lukechilds) [inline apps/registry.json:801]: We should set `torOnly` to `true` since currently the app won't work when hosted from an `*.onion` hostname.

```suggestion
        "deterministicPassword": true,
        "torOnly": true
```

**Comment by lukechilds** [comment]
When the user opens an order, that sends to message to the `ws` service right? How does the `backend` service know to pay the invoice? How are `backend` and `ws` talking to each other?

**Comment by lukechilds** [comment]
Oh also one more thing, plausible.io seems like a good and privacy respecting analytics service, but we should probably notify users of this somehow. I think just simply adding some text the app store description stating that the app collects anonymised usage statistics should be sufficient.

**Comment by djkazic** [comment]
1. We will remove plausible, we had this in the codebase to help with debugging in the beginning. I'll let you know once we have an updated build without it.
2. `backend` and `ws` talk to each other over zmq. I'll link you the lines of code responsible:
https://github.com/kolliderhq/umbrel-server/blob/5531c670fee73e31bd2d6849c94d9b0b33bac26b/backend/lnhedgehog.py#L22-L23
https://github.com/kolliderhq/umbrel-server/blob/486cfcb822aa36c05a670ef3dc809c37dbfb6876/ws-server/ws_server.js#L21-L31

**Comment by djkazic** [comment]
Also, I'm a bit confused, the flag `torOnly` would suggest that we want to set it to false, right? Just checking -- I'll make the change ASAP once I understand how it's used.

**Comment by lukechilds** [comment]
>Also, I'm a bit confused, the flag torOnly would suggest that we want to set it to false, right? Just checking -- I'll make the change ASAP once I understand how it's used.

Oh yes you're completely right! Sorry, the usecase for that flag was for apps that are only accessible over Tor. We have the opposite issue here, we only want to expose locally. I just got confused.

Your app gets passed an `$APP_HIDDEN_SERVICE` env var which is the address of the Tor hidden servivce it will be served at. Could we use this to get Tor working? That won't help us right cos we still need to add the CORS header to your remote server? Is the hostname check on the local app or the remote server?

If it's just on the remote server could we run a local proxy to resolve the issue? If it's on the local app, what is it protecting against, wouldn't the browser's default CORS policy provide sufficient protection?

**Comment by lukechilds** [comment]
>backend and ws talk to each other over zmq. I'll link you the lines of code responsible:
> https://github.com/kolliderhq/umbrel-server/blob/5531c670fee73e31bd2d6849c94d9b0b33bac26b/backend/lnhedgehog.py#L22-L23
> https://github.com/kolliderhq/umbrel-server/blob/486cfcb822aa36c05a670ef3dc809c37dbfb6876/ws-server/ws_server.js#L21-L31

Thanks! Looks like this repo is private, are you planning to make the source available publically?

**Comment by djkazic** [comment]
It won't work right now because of the remote server only whitelisting umbrel.local. Might be something we can ship in the next release with more testing. I've set `torOnly=false`.

**Comment by lukechilds** [comment]
Got it. No worries. Maybe for now we could just handle the error in the UI, something simple like:

```
if (hostname !== 'umbrel.local') {
  alert("Kollider only currently supports acces from umbrel.local");
}
```

Just so Tor users aren't left confused when it doesn't work. What do you think?

**Comment by djkazic** [comment]
I think that's a reasonable change. I'll make the repository public as well, just need to ping my teammate.

**Comment by lukechilds** [comment]
Ahh just seen in your source files you're hardcoding the Umbrel IPs. Ideally these would be passed in as environment variables.

So something like this in `umbrel-server`:

```python
ZMQ_ADDRESS = os.environ.get("KOLLIDER_ZMQ_ADDRESS");
ZMQ_SUB_ADDRESS = os.environ.get("KOLLIDER_ZMQ_SUB_ADDRESS");
ZMQ_HEDGER_ADDRESS = os.environ.get("KOLLIDER_ZMQ_HEDGER_ADDRESS");
ZMQ_HEDGER_SUB_ADDRESS = os.environ.get("KOLLIDER_HEDGER_SUB_ADDRESS");
```

and then passed through in the Umbrel app like:

```yaml
  ws:
    image: kolliderhq/kollider-ws-client:v1.0.0
    init: true
    user: 1000:1000
    restart: on-failure
    stop_grace_period: 1m
    ports:
      - "$APP_KOLLIDER_WS_PORT:8080"
    environment:
      APP_PASSWORD: $APP_PASSWORD
      KOLLIDER_ZMQ_ADDRESS: "tcp://${APP_KOLLIDER_HH_IP}:5556";
      KOLLIDER_ZMQ_SUB_ADDRESS: "tcp://${APP_KOLLIDER_HH_IP}:5557";
      KOLLIDER_ZMQ_HEDGER_ADDRESS: "tcp://${APP_KOLLIDER_HH_IP}:5558";
      KOLLIDER_HEDGER_SUB_ADDRESS: "tcp://${APP_KOLLIDER_HH_IP}:5559";
    networks:
      default:
        ipv4_address: $APP_KOLLIDER_WS_IP
```

The IPs aren't guaranteed to always be the same. They could change in the future. e.g there's currently a collission with the IPs used in this PR and the Suredbits Wallet PR branch: https://github.com/getumbrel/umbrel/pull/1147/files#diff-ecb3d5615afcbe1e5df55695252c8f0835e8cb1ec10a6048a83cf0a80d2a1ab3R226-R229. If we changed your IPs in the configure script it would break your `umbrel-server` if it's hardcoded.

This isn't super high priority if it's gonna be an issue getting it done in time for this release. I can just make sure we don't change them for now. But they would ideally be updated to work like this at some point to prevent breakage in the future. Same logic applies if you've hardcoded IPs anywhere else.

>we have a small alert that fires when you access it from non umbrel.local origins.

Awesome! Testing now!

**Comment by djkazic** [comment]
@lukechilds I've made the requested changes for IP addresses and docker-compose.

**Comment by lukechilds** [comment]
Tested and working great!

Are you able to merge these tweaks in and then this is good for merge: https://github.com/kolliderhq/umbrel/pull/2

**Comment by lukechilds** [comment]
Nw added my changes on top here to get this merged quickly: https://github.com/getumbrel/umbrel/pull/1221

Thanks so much for working on this, can't wait to ship it!

**Comment by mayankchhabra** [comment]
Congrats on the launch, guys! The app's really well done. Great to see users already opening positions and enjoying the app :)

[Here](https://github.com/getumbrel/umbrel-apps-gallery/tree/master/kollider) are the app store gallery images and app icon for Kollider. If you'd like to update them in the future, feel free to directly open a PR at [getumbrel/umbrel-apps-gallery](https://github.com/getumbrel/umbrel-apps-gallery).

Cheers!

**Pull request closed without merging**
